syntax = "proto3";

package foreverbull.backtest;

import "google/protobuf/timestamp.proto";
import "foreverbull/finance/finance.proto";
import "foreverbull/common.proto";

option go_package = "github.com/lhjnilsson/foreverbull/pkg/backtest/pb";

message Execution {
    message Status {
        enum Status {
            CREATED = 0;
            RUNNING = 1;
            COMPLETED = 2;
            FAILED = 3;
        }
        Status status = 1;
        optional string error = 2;
        google.protobuf.Timestamp occurred_at = 3;
    }
    string id = 1;
    string session = 2;
    foreverbull.common.Date start_date = 3;
    foreverbull.common.Date end_date = 4;
    optional string benchmark = 5;
    repeated string symbols = 6;
    repeated Status statuses = 7;
}

message Period {
    foreverbull.common.Date date = 1;
    double PNL = 2;
    double returns = 3;
    double portfolio_value = 4;
    int32 longs_count = 5;
    int32 shorts_count = 6;
    double long_value = 7;
    double short_value = 8;
    double starting_exposure = 9;
    double ending_exposure = 10;
    double long_exposure = 11;
    double short_exposure = 12;
    double capital_used = 13;
    double gross_leverage = 14;
    double net_leverage = 15;
    double starting_value = 16;
    double ending_value = 17;
    double starting_cash = 18;
    double ending_cash = 19;
    double max_drawdown = 20;
    double max_leverage = 21;
    double excess_return = 22;
    double treasury_period_return = 23;
    double algorithm_period_return = 24;
    optional double algo_volatility = 25;
    optional double sharpe = 26;
    optional double sortino = 27;
    optional double benchmark_period_return = 28;
    optional double benchmark_volatility = 29;
    optional double alpha = 30;
    optional double beta = 31;

    repeated foreverbull.finance.Position positions = 32;
}
